modules/client/src/internal/utils.ts

```diff
@@ -824,3 +863,801 @@ export function toPluginPermissionOperationType(
+          break;
+        default:
+          classifiedActions.push(ProposalActionTypes.ACTION_NOT_ALLOWED);
+          break;
+      }
+    } catch {
+      classifiedActions.push(ProposalActionTypes.UNKNOWN);
+    }
+  }
+  return classifiedActions;
+}
+
+/**
+ * Returns true if the actions are valid for a plugin update proposal with root permission
+ *
+ * @export
+ * @param {ProposalActionTypes[]} actions
+ * @return {*}  {boolean}
+ */
+export function isPluginUpdateActionWithRootPermission(
+  actions: ProposalActionTypes[],
+): boolean {
+  // get the first 5 actions
+  // if the first is a dao upgrade skip it
+  let receivedPattern;
+  if (
+    actions[0] === ProposalActionTypes.UPGRADE_TO ||
+    actions[0] === ProposalActionTypes.UPGRADE_TO_AND_CALL
+  ) {
+    receivedPattern = actions.slice(1, 6);
+  } else {
+    receivedPattern = actions.slice(0, 5);
+  }
+  // check if it matches the expected pattern
+  // length should be 5
+  if (
+    receivedPattern.length !== 5 ||
+    !compareArrays(receivedPattern, PLUGIN_UPDATE_WITH_ROOT_ACTION_PATTERN)
+  ) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Returns true if the actions are valid for a plugin update proposal without root permission
+ *
+ * @export
+ * @param {ProposalActionTypes[]} actions
+ * @return {*}  {boolean}
+ */
+export function isPluginUpdateAction(actions: ProposalActionTypes[]): boolean {
+  // get the first 3 action
+  // if the first is a dao upgrade skip it
+  let receivedPattern;
+  if (
+    actions[0] === ProposalActionTypes.UPGRADE_TO ||
+    actions[0] === ProposalActionTypes.UPGRADE_TO_AND_CALL
+  ) {
+    receivedPattern = actions.slice(1, 4);
+  } else {
+    receivedPattern = actions.slice(0, 3);
+  }
+  // check if it matches the expected pattern
+  // length should be 3
+  if (
+    receivedPattern.length !== 3 ||
+    !compareArrays(receivedPattern, PLUGIN_UPDATE_ACTION_PATTERN)
+  ) {
+    return false;
+  }
+  return true;
+}
+/**
+ * Returns true if the actions are valid for a plugin update proposal
+ *
+ * @export
+ * @param {ProposalActionTypes[]} actions
+ * @return {*}  {boolean}
+ */
+export function isDaoUpdateAction(actions: ProposalActionTypes[]): boolean {
+  // UpgradeTo or UpgradeToAndCall should be the first action
+  return actions[0] === ProposalActionTypes.UPGRADE_TO ||
+    actions[0] === ProposalActionTypes.UPGRADE_TO_AND_CALL;
+}
+
+export function validateUpdateDaoProposalActions(
+  actions: DaoAction[],
+  daoAddress: string,
+  implementationAddress: string,
+  currentDaoVersion: [number, number, number],
+): DaoUpdateProposalValidity {
+  const classifiedActions = classifyProposalActions(actions);
+  const causes: DaoUpdateProposalInvalidityCause[] = [];
+  // check if the actions are valid
+  // if they are not valid return add
+  // invalid actions to the causes and return
+  if (isDaoUpdateAction(classifiedActions)) {
+    // if they are valid, the upgrade action must
+    // be the first one
```

Comment:
That's hacky and implict

Reply:
No, is what we talked about, the DAO update should always bee the first thing in case it exists. The app will need at least wo functioons, one for checking if is a dao update and another for checking if it is a proposal update, because they can do ir separately